hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java

```diff
@@ -81,7 +99,55 @@ public String getPartitionPath(Row row) {
     return getKey(genericRecord).getPartitionPath();
   }
 
-  void buildFieldPositionMapIfNeeded(StructType structType) {
+  /**
+   * Fetch partition path from {@link InternalRow}.
+   *
+   * @param internalRow {@link InternalRow} instance from which partition path needs to be fetched from.
+   * @param structType  schema of the internalRow.
+   * @return the partition path.
+   */
+  public String getPartitionPath(InternalRow internalRow, StructType structType) {
+    Row row = null;
+    try {
+      row = deserializeRow(getEncoder(structType), internalRow);
+    } catch (Exception e) {
+      throw new IllegalStateException("Convertion of InternalRow to Row failed with exception " + e);
+    }
+    return getPartitionPath(row);
+  }
+
+  private ExpressionEncoder getEncoder(StructType structType) {
+    if (encoder == null) {
+      synchronized (this) {
+        encoder = getRowEncoder(structType);
+      }
+    }
+    return encoder;
+  }
+
+  private static ExpressionEncoder getRowEncoder(StructType schema) {
+    List<Attribute> attributes = JavaConversions.asJavaCollection(schema.toAttributes()).stream()
+        .map(Attribute::toAttribute).collect(Collectors.toList());
+    return RowEncoder.apply(schema)
+        .resolveAndBind(JavaConverters.asScalaBufferConverter(attributes).asScala().toSeq(),
+            SimpleAnalyzer$.MODULE$);
+  }
+
+  private static Row deserializeRow(ExpressionEncoder encoder, InternalRow row)
```

Comment:
yet to test this method. I found a similar method for serializeFromRow and came up with this. But have fixed all other build in key gens like simple, complex, timestamp, custom. Will update the patch once I have the fix. Wanted to open up for reviews as I work on them.